docs/proposals/role_decomposition.md

```diff
@@ -0,0 +1,322 @@
+# Scaffolding for decomposing large roles
+
+## Why?
+
+Currently we have roles that are very large and encompass a lot of different
+components. This makes for a lot of logic required within the role, can
+create complex conditionals, and increases the learning curve for the role.
+
+## How?
+
+Creating a guide on how to approach breaking up a large role into smaller,
+component based, roles. Also describe how to develop new roles, to avoid creating
+large roles.
+
+## Proposal
+
+Create a new guide or append to the current contributing guide a process for
+identifying large roles that can be split up, and how to compose smaller roles
+going forward.
+
+### Large roles
+
+A role should be considered for decomposition if it:
+
+1) Configures/installs more than one product.
+1) Can configure multiple variations of the same product that can live
+side by side.
+1) Has different entry points for upgrading and installing a product
+
+Large roles should be responsible for:
+
+1) Composing smaller roles to provide a full solution such as an Openshift Master
```

Comment:
On one hand I like this idea. It will help to decrease redundancy and role complexity. On the other hand, instead of one `openshift-master` role, we will have:
- `openshift-master-configure` (responsible for various configs, certs, ...) 
- `openshift-master-system-services` (responsible for systemd services a.k.a pulling and running images) 
- `openshift-master-restart` (responsible for restarting and waiting for services/nodes to come back/alive)
- `openshift-master-rest` (responsible for the rest of the tasks)

This is just an illustrative example, not saying it needs to be done this way. Just have the master installation/upgrade tasks in mind as there is a lot of tasks common.